### ISSUE TITLE
Smaller icons

### ISSUE THREAD

**Issue by bhollis** [Bungie-net/api#640]
For DIM, we'd love the ability to use smaller versions of the item icons, so that we use less memory when displaying a lot of tiny icons. 1/2, 1/4, 1/8 sizes, etc (an image pyramid) would be ideal!

**Comment by floatingatoll** [comment]
Which browser are you using that is showing high memory usage for images?
Safari macOS reports that about half (200MB of 450MB footprint) RAM used on
beta is JavaScript objects without any RAM particularly being used by
images.

**Comment by bhollis** [comment]
While it's possible that the browser is doing a good job of throwing away the source images after resize, it still needs to decode and resize a large image in order to use it in a small space, which is wasteful.

**Comment by vthornheart-bng** [comment]
Seems like a reasonable request, though I will give the caveat that I don't necessarily know when we're going to get to it.  Part of me would like, if we were to do that, to refactor how we return image data in some way such that we reduce all of the duplication of things like the relative path.  (though to do that would break backcompat, so I'll resist that for now.)

**Comment by vthornheart-bng** [comment]
Okay, here's the gameplan!

There's going to be a new property in the manifest, "iconImagePyramidInfo" - this will be a list of objects that contain:

* the "name" of the set of resized images - and also the subfolder under the /icons/ directory where the resized icons will be located
* the "factor" by which the width and height of the existing image is multiplied to shrink the image.  Half size will be .5, quarter size .25, 1/8th size will be .125.  Those are the three we're planning to support for now, but I'm building it such that we can change or add to it if needed.

We will not return the full path to each version of the image, as I don't want to increase the size of the definitions with redundancies any further than they already are.  To find the resized icons, you'll need to grab the "name" property of the size you care about, and use that as the name of a subfolder under where you find the original icon that *is* defined.  The name of the resized icon will be identical to the name of the original icon.

I'm working through some of the implications and changes that would need to be made in order to make this work seamlessly as we speak.  I'll let you know once I have it in place and we know it'll be ready for a future deployment.

In the meantime, let me know if you have any questions or concerns.  I know that not putting all versions in the definition is going to make them a bit less discoverable, but we need to take the tradeoff for reduced definition size and redundancy.  Hopefully this will be an acceptable compromise.

**Comment by bhollis** [comment]
That sounds pretty good! Let me think about how I’d write a component to take advantage of that. 

One thing that’d really help there is to return the asset dimensions- then I can write code to pick the right asset version based on its rendered size and the asset size.

**Comment by floatingatoll** [comment]
Would returning the 1.0x dimensions in iconImagePyramidInfo be sufficient,
or are you hoping to see the dimensions listed for all four sizes?

**Comment by vthornheart-bng** [comment]
Hmm, good question.  The dimensions of the asset will vary depending on the specific image we're looking at - so if I give you a half size Emblem image, for example, it'll be different dimensions than an item icon.  I could see returning the original icon's size potentially as a property alongside the icon itself, since I can't trust that they won't be different.  But I also don't want to create a great deal of redundancy.

Perhaps the best thing might be for me to document the known/expected sizes for properties that consistently have a given expected size in the documentation, rather than returning that as actual data in the definitions.  Then when you use an icon, you can determine which size you want to use based on the documentation for that icon property and where you're using it in your UI.

... I've also got to solve the problem of icons that are added through other sources, which I just realized as I was talking about this.  The nice, clean automated solution I mentioned above falls apart for the various places where icons aren't sourced via direct game content relationships.  This could get uglier than I had previously accounted for, as usual. :)

**Comment by floatingatoll** [comment]
This is a scenario where a 50% or 80% fix would still help, so don't worry
if you ship it with "only some assets are covered, no promises" :)

**Comment by vthornheart-bng** [comment]
Perhaps, but I'd hate it if you ended up having to have some kind of special handling particularly when we source the same property (like displayProperties.icon) differently for different types of entities, and some were handled where others weren't.  That would suck.  I'd like to make sure even if we don't have all properties map in this way, that the ones that do consistently get these resized images.  I think I might have to add a post-build step to do this, which I was hoping to avoid, but will probably be for the best to avoid that situation. :(

**Comment by ckhicks** [comment]
Would including the dimensions for the original asset suffice, letting clients perform the math on their side and use the proper path?

**Comment by ckhicks** [comment]
Take this with a grain of salt, as this doesn't directly affect anything of mine. XD

I wonder if, similar to @floatingatoll's suggestion, including *only* a single 1.0x base resolution would suffice, so long as the clients knew the available multipliers. While that may mean that I have two sources with varying original resolutions, I could at least parse those values in my app and determine whether or not to use the 25% or 50% variant, for example.

I also assume there will be cases where the smallest resolution isn't provided (or is duplicated) for assets that are already really small, ya?

Hopefully I'm not over-simplifying it for someone else's use case, but that seems reasonable and keeps your computation down at build time.

**Comment by vthornheart-bng** [comment]
Yeah, I dig that.  And I'd actually be providing the smallest resolution for everything, as I wanted it to be fully automated and didn't want to run into a scenario where people had to guess which sizes would be there... I can see where some of them might end up being uselessly small as a result though.  But hopefully if you need it for a use case where you know the icons being returned are very small, you'd make choices about the size you use based on that knowledge.

This discussion definitely makes me wish I could turn back time and make every icon property an object rather than a mere string path.  I'd like to add more metadata to each of them (perhaps even just the "names" of the resized icons they have, and the size of the largest one) without it being a bunch of one-off properties that you have to know are associated with one-another.  That'd definitely be an improvement that I could see happening for the next major revision of the API, if and when that day ever comes.

**Comment by bhollis** [comment]
Yeah just having dimensions for the 1.0x version is fine, I can calculate the rest. Or a documented set of initial sizes.

**Comment by vthornheart-bng** [comment]
Okay, time is coming short so it's time for me to take a pragmatic pivot.

For now at least, I'm going to make this resizing *just* happen to icons pointed at by the displayProperties.icons property, rather than all icons/images in Destiny Definitions universally.  Ideally that should cover the majority of use cases.  If we ever get the time to do so, I can loop back and make a more universal solution that applies to everyone.  (I hesitate to do something like create a file handler that dynamically generates these on the fly, which would solve the problem in a centralized way but might tax our servers, particularly when new content is released.  I'd much rather generate these once at build time and deploy them as static assets, but it does create some unpleasant implications for parts of the import and definitions that were never built in anticipation of us wanting to provide multiple sizes of images).

Anyways, I digress - that's the situation right now.  Let me know if that will be sufficient - I imagine it will be if the problem you all are running into is related to trying to load many DestinyInventoryItemDefinition icons at once.

**Comment by bhollis** [comment]
I think that'd work fine!

**Comment by vthornheart-bng** [comment]
Sweet, good deal!
